pilot/pkg/networking/plugin/authz/rbac.go

```diff
@@ -166,36 +167,65 @@ func generateMetadataStringMatcher(keys []string, v *metadata.StringMatcher) *me
 	return &metadata.MetadataMatcher{
 		Filter: authn.AuthnFilterName,
 		Path:   paths,
-		Value: &metadata.MetadataMatcher_Value{
-			MatchPattern: &metadata.MetadataMatcher_Value_StringMatch{
+		Value: &metadata.ValueMatcher{
+			MatchPattern: &metadata.ValueMatcher_StringMatch{
 				StringMatch: v,
 			},
 		},
 	}
 }
 
-// createDynamicMetadataMatcher creates a MetadataMatcher for the given key, value pair.
-func createDynamicMetadataMatcher(k, v string) *metadata.MetadataMatcher {
-	var keys []string
+func generateMetadataListMatcher(keys []string, v string) *metadata.MetadataMatcher {
 	forceRegexPattern := false
+	listMatcher := &metadata.ListMatcher{
+		MatchPattern: &metadata.ListMatcher_OneOf{
+			OneOf: &metadata.ValueMatcher{
+				MatchPattern: &metadata.ValueMatcher_StringMatch{
+					StringMatch: createStringMatcher(v, forceRegexPattern),
+				},
+			},
+		},
+	}
+
+	paths := make([]*metadata.MetadataMatcher_PathSegment, 0)
+	for _, k := range keys {
+		paths = append(paths, &metadata.MetadataMatcher_PathSegment{
+			Segment: &metadata.MetadataMatcher_PathSegment_Key{Key: k},
+		})
+	}
+
+	return &metadata.MetadataMatcher{
+		Filter: authn.AuthnFilterName,
+		Path:   paths,
+		Value: &metadata.ValueMatcher{
+			MatchPattern: &metadata.ValueMatcher_ListMatch{
+				ListMatch: listMatcher,
+			},
+		},
+	}
+}
+
+func generateKeys(k string) ([]string, error) {
```

Comment:
Can you use more descriptive name, e.g MetadataKey or smt that self-explain what the key is for.

Reply:
Done.